parser/src/grammar_builder.rs

```diff
@@ -296,10 +298,50 @@ impl GrammarBuilder {
             }
         }
 
+        let token_ranges = if negate {
+            let trie = if let Some(trie) = &trie {
+                trie
+            } else {
+                bail!("negation of token ranges is not supported without a tokenizer");
```

Comment:
No tokenizer is only possible if we're just validating the grammar, not actually compiling it. You can do the same as `special_token` below - return `INVALID_TOKEN` or similar and add warning